src/gui/qgsmetadatawidget.cpp

```diff
@@ -0,0 +1,708 @@
+  mDefaultCategories << tr( "Utilities Communication" ) << tr( "Economy" ) << tr( "Society" ) << tr( "Imagery Base Maps Earth Cover" );
+  mDefaultCategoriesModel = new QStringListModel( mDefaultCategories );
+  mDefaultCategoriesModel->sort( 0 );  // Sorting using translations
+  listDefaultCategories->setModel( mDefaultCategoriesModel );
+
+  // Categories
+  mCategoriesModel = new QStringListModel();
+  listCategories->setModel( mCategoriesModel );
+
+  // Rights
+  mRightsModel = new QStringListModel();
+  listRights->setModel( mRightsModel );
+
+  tabWidget->setCurrentIndex( 0 );
+
+  // Setup the link view
+  mLinksModel = new QStandardItemModel();
+  mLinksModel->setColumnCount( 7 );
+  QStringList headers = QStringList();
+  headers << tr( "Name" ) << tr( "Type" ) << tr( "URL" ) << tr( "Description" ) << tr( "Format" ) << tr( "MIME" ) << tr( "Size" );
+  mLinksModel->setHorizontalHeaderLabels( headers );
+  tabLinks->setModel( mLinksModel );
+  tabLinks->setItemDelegate( new LinkItemDelegate() );
+
+  connect( tabWidget, &QTabWidget::currentChanged, this, &QgsMetadataWidget::updatePanel );
+  connect( btnAutoSource, &QPushButton::clicked, this, &QgsMetadataWidget::setAutoSource );
+  connect( btnAddVocabulary, &QPushButton::clicked, this, &QgsMetadataWidget::addVocabulary );
+  connect( btnRemoveVocabulary, &QPushButton::clicked, this, &QgsMetadataWidget::removeVocabulary );
+  connect( btnAddRight, &QPushButton::clicked, this, &QgsMetadataWidget::addRight );
+  connect( btnRemoveRight, &QPushButton::clicked, this, &QgsMetadataWidget::removeRight );
+  connect( btnAddLicence, &QPushButton::clicked, this, &QgsMetadataWidget::addLicence );
+  connect( btnRemoveLicence, &QPushButton::clicked, this, &QgsMetadataWidget::removeLicence );
+  connect( btnAutoCrs, &QPushButton::clicked, this, &QgsMetadataWidget::setAutoCrs );
+  connect( btnAddContact, &QPushButton::clicked, this, &QgsMetadataWidget::addContact );
+  connect( btnRemoveContact, &QPushButton::clicked, this, &QgsMetadataWidget::removeContact );
+  connect( tabContacts, &QTableWidget::itemSelectionChanged, this, &QgsMetadataWidget::updateContactDetails );
+  connect( btnAddLink, &QPushButton::clicked, this, &QgsMetadataWidget::addLink );
+  connect( btnRemoveLink, &QPushButton::clicked, this, &QgsMetadataWidget::removeLink );
+  connect( btnNewCategory, &QPushButton::clicked, this, &QgsMetadataWidget::addNewCategory );
+  connect( btnAddDefaultCategory, &QPushButton::clicked, this, &QgsMetadataWidget::addDefaultCategory );
+  connect( btnRemoveCategory, &QPushButton::clicked, this, &QgsMetadataWidget::removeCategory );
+
+  fillComboBox();
+  setPropertiesFromLayer();
+  updateContactDetails();
+}
+
+QgsMetadataWidget::~QgsMetadataWidget()
+{
+}
+
+void QgsMetadataWidget::setAutoSource() const
```

Comment:
Setters are normally not const, is this line correct?

Reply:
Sorry, it's misleading, it's not a setter. It's a function to set text automatically in a QLineEdit according to the layer
![screen shot 2017-09-14 at 11 28 49](https://user-images.githubusercontent.com/1609292/30422341-eb33b44c-993f-11e7-9eb4-154d6cfbf9c3.png)